contracts/data-storage/andromeda-graph/src/contract.rs

```diff
@@ -0,0 +1,395 @@
+    GetDataOwnerResponse, PointCoordinate, QueryMsg as PointQueryMsg,
+};
+use andromeda_std::{
+    ado_base::{InstantiateMsg as BaseInstantiateMsg, MigrateMsg},
+    ado_contract::ADOContract,
+    amp::AndrAddr,
+    common::{actions::call_action, context::ExecuteContext, encode_binary},
+    error::ContractError,
+    os::aos_querier::AOSQuerier,
+};
+
+use crate::state::{MAP_INFO, MAP_POINT_INFO, TOTAL_POINTS_NUMBER, USER_COORDINATE};
+
+// version info for migration info
+const CONTRACT_NAME: &str = "crates.io:andromeda-graph";
+const CONTRACT_VERSION: &str = env!("CARGO_PKG_VERSION");
+
+#[cfg_attr(not(feature = "library"), entry_point)]
+pub fn instantiate(
+    deps: DepsMut,
+    env: Env,
+    info: MessageInfo,
+    msg: InstantiateMsg,
+) -> Result<Response, ContractError> {
+    let resp = ADOContract::default().instantiate(
+        deps.storage,
+        env,
+        deps.api,
+        &deps.querier,
+        info,
+        BaseInstantiateMsg {
+            ado_type: CONTRACT_NAME.to_string(),
+            ado_version: CONTRACT_VERSION.to_string(),
+            kernel_address: msg.kernel_address,
+            owner: msg.owner,
+        },
+    )?;
+
+    MAP_INFO.save(deps.storage, &msg.map_info)?;
+    TOTAL_POINTS_NUMBER.save(deps.storage, &0)?;
+
+    Ok(resp)
+}
+
+#[cfg_attr(not(feature = "library"), entry_point)]
+pub fn execute(
+    deps: DepsMut,
+    env: Env,
+    info: MessageInfo,
+    msg: ExecuteMsg,
+) -> Result<Response, ContractError> {
+    let ctx = ExecuteContext::new(deps, info, env);
+    match msg {
+        ExecuteMsg::AMPReceive(pkt) => {
+            ADOContract::default().execute_amp_receive(ctx, pkt, handle_execute)
+        }
+        _ => handle_execute(ctx, msg),
+    }
+}
+
+fn handle_execute(mut ctx: ExecuteContext, msg: ExecuteMsg) -> Result<Response, ContractError> {
+    let action = msg.as_ref().to_string();
+
+    let action_response = call_action(
+        &mut ctx.deps,
+        &ctx.info,
+        &ctx.env,
+        &ctx.amp_ctx,
+        msg.as_ref(),
+    )?;
+
+    let res = match msg {
+        ExecuteMsg::UpdateMap { map_info } => execute_update_map(ctx, map_info, action),
+        ExecuteMsg::StoreCoordinate {
+            coordinate,
+            is_timestamp_allowed,
+        } => execute_store_coordinate(ctx, coordinate, is_timestamp_allowed, action),
+        ExecuteMsg::StoreUserCoordinate {
+            user_location_paths,
+        } => execute_store_user_coordinate(ctx, user_location_paths, action),
+        ExecuteMsg::DeleteUserCoordinate { user } => {
+            execute_delete_user_coordinate(ctx, user, action)
+        }
+        _ => ADOContract::default().execute(ctx, msg),
+    }?;
+
+    Ok(res
+        .add_submessages(action_response.messages)
+        .add_attributes(action_response.attributes)
+        .add_events(action_response.events))
+}
+
+pub fn execute_update_map(
+    ctx: ExecuteContext,
+    map_info: MapInfo,
+    action: String,
```

Comment:
I'm not sure this parameter is necessary. The action is static so can be hardcoded in the attributes.

Reply:
Removed this parameter.